src/patterns.rst

```diff
@@ -639,333 +638,321 @@ Struct Patterns
+:dp:`fls_3vgmkm2mzwwy`
+The :t:`type` of a :t:`matched shorthand deconstructor` and the :t:`type` of the
+matched :t:`field` shall be :t:`unifiable`.
 
-* :dp:`fls_rspzc5jqbysa`
-  Contain a :t:`matched shorthand deconstructor`, or
+:dp:`fls_m91ith3rjy79`
+If the :t:`deconstructee` of a :t:`record struct pattern` is a :t:`record enum
+variant` or a :t:`record struct`, then
 
-* :dp:`fls_2l9wbc8sqtlo`
-  Has a :s:`RecordStructRestPattern` or a
-  :s:`RecordStructFieldDeconstructorList` with a :s:`RecordStructRestPattern`.
+* :dp:`fls_c09jf2vpcr58`
+  For each :t:`field` of the :t:`deconstructee`, the :t:`record struct pattern`
+  shall either:
 
-.. rubric:: Examples
+  * :dp:`fls_4h00oqypa8qg`
+    Contain at most one :t:`matched named deconstructor`, or
 
-.. syntax::
+  * :dp:`fls_195mqijyrnam`
+    Contain at most one :t:`matched shorthand deconstructor`, or
 
-   struct Struct {
-   	field: u32,
-   	other: u32,
-   }
+  * :dp:`fls_ta0vdoqmt2k1`
+    Have exactly one :s:`RecordStructRestPattern`.
 
-   let Struct { field, other };
-   let Struct { field, .. };
-   let Struct { .. };
+* :dp:`fls_f0u0j4q90lpl`
+  A :s:`RecordStructRestPattern` is allowed even if all :t:`[field]s` of the
+  :t:`deconstructee` have been matched.
 
-.. _fls_cbj864ya6vli:
+:dp:`fls_8bi8q3usubby`
+If the :t:`deconstructee` of a :t:`record struct pattern` is a :t:`tuple enum
+variant` or a :t:`tuple struct`, then
 
-Tuple Struct Patterns
-^^^^^^^^^^^^^^^^^^^^^
+* :dp:`fls_1x0o71kxj3yq`
+  For each :t:`field` of the :t:`deconstructee`, the :t:`record struct pattern`
+  shall either:
 
-.. rubric:: Legality Rules
+  * :dp:`fls_1thgpx95lfg5`
+    Contain at most one :t:`matched index deconstructor`, or
 
-:dp:`fls_ec0o57hwg8ic`
-The :t:`deconstruction type` of a :t:`tuple struct pattern` shall resolve to a
-:t:`tuple struct type`.
+  * :dp:`fls_rpo1wimbmzhc`
+    Have exactly one ``RecordStructRestPattern.``
 
-:dp:`fls_3e2zaeqo1s96`
-A :t:`tuple struct pattern` shall contain one :t:`subpattern` for each
-:t:`field` of the :t:`deconstruction type`.
+* :dp:`fls_brhtaaxt1s3s`
+  A :s:`RecordStructRestPattern` is allowed even if all :t:`[field]s` of the
+  :t:`deconstructee` have been matched.
 
-:dp:`fls_w936pvga6lgn`
-A :t:`subpattern` of a :t:`tuple struct pattern` matches a :t:`field` of the
-:t:`deconstruction type` when its position and the position of the :t:`field`
-in the :t:`deconstruction type` are the same. Such a :t:`subpattern` is a
-:dt:`matched subpattern`.
+:dp:`fls_jwz3arnfkxwn`
+If the :t:`deconstructee` of a :t:`record struct pattern` is a :t:`union`, then
 
-:dp:`fls_aeh8bzh59m05`
-The :t:`type` of the :t:`subpattern` of a :t:`tuple struct pattern` and the
-:t:`type` of the matched :t:`field` shall be :t:`unifiable`.
+* :dp:`fls_pfz8xlwezbw1`
+  The :s:`RecordStructPatternContent` of the :t:`record struct
+  pattern` shall contain exactly one :s:`FieldDeconstructor` and no
+  :s:`RecordStructRestPattern`.
```

Comment:
If we want to keep this 100% symmetric with Struct Expressions, we can reword this as

The :t:`record struct pattern` shall not contain a :s:`RecordStructRestPattern`.